File: typescript/protocol/protocol-adapter.ts
import type { Transport } from './transport.js';
import { CDPSession } from './cdp-session.js';
import { BiDiSession } from './bidi-session.js';
import { SpiderEventEmitter } from '../events/emitter.js';
import { getKeyParams } from './types.js';
import { logger } from '../utils/logger.js';

/**
 * Unified protocol interface that wraps either CDPSession or BiDiSession
 * based on the browser type.
 *
 * - chrome, chrome-new, chrome-h, servo, lightpanda → CDP
 * - firefox → BiDi
 */
export interface ProtocolAdapterOptions {
  commandTimeoutMs?: number;
}

export class ProtocolAdapter {
  private cdp: CDPSession | null = null;
  private bidi: BiDiSession | null = null;
  private transport: Transport;
  private emitter: SpiderEventEmitter;
  private protocol: 'cdp' | 'bidi' | 'auto';
  private adapterOpts: ProtocolAdapterOptions;

  constructor(transport: Transport, emitter: SpiderEventEmitter, browser: string, opts?: ProtocolAdapterOptions) {
    this.transport = transport;
    this.emitter = emitter;
    this.adapterOpts = opts ?? {};

    const sessionOpts = opts?.commandTimeoutMs ? { commandTimeoutMs: opts.commandTimeoutMs } : undefined;

    // Determine protocol from browser type
    if (browser === 'auto') {
      this.protocol = 'auto';
      // Don't create sessions yet — auto-detect in init()
    } else if (browser === 'firefox') {
      this.protocol = 'bidi';
      this.bidi = new BiDiSession(transport, sessionOpts);
    } else {
      this.protocol = 'cdp';
      this.cdp = new CDPSession(transport, sessionOpts);
    }

    // Wire up message routing
    transport.onMessage((data: string) => this.routeMessage(data));
  }

  /** Whether this adapter uses CDP or BiDi (or 'auto' before init). */
  get protocolType(): 'cdp' | 'bidi' | 'auto' {
    return this.protocol;
  }

  /** Route incoming WebSocket messages to the right session + spider events. */
  private routeMessage(data: string): void {
    // Check for Spider.* events first
    try {
      const msg = JSON.parse(data);
      if (typeof msg.method === 'string' && msg.method.startsWith('Spider.')) {
        this.handleSpiderEvent(msg.method, msg.params ?? {});
        return;
      }
    } catch {
      // ignore parse errors for spider event check
    }

    // Route to protocol session
    if (this.cdp) {
      this.cdp.handleMessage(data);
    } else if (this.bidi) {
      this.bidi.handleMessage(data);
    }
  }

  /** Handle Spider.* custom events from the server. */
  private handleSpiderEvent(method: string, params: Record<string, unknown>): void {
    switch (method) {
      case 'Spider.captchaDetected':
        this.emitter.emit('captcha.detected', {
          types: (params.types as string[]) ?? [],
          url: (params.url as string) ?? '',
        });
        break;
      case 'Spider.captchaSolving':
        this.emitter.emit('captcha.solving', {
          types: (params.types as string[]) ?? [],
          url: (params.url as string) ?? '',
          round: (params.round as number) ?? 0,
        });
        break;
      case 'Spider.captchaSolved':
        this.emitter.emit('captcha.solved', { url: (params.url as string) ?? '' });
        break;
      case 'Spider.captchaFailed':
        this.emitter.emit('captcha.failed', {
          url: (params.url as string) ?? '',
          reason: (params.reason as string) ?? '',
        });
        break;
      default:
        logger.debug(`unhandled Spider event: ${method}`, params);
    }
  }

  // -------------------------------------------------------------------
  // Unified interface — all methods work for both CDP and BiDi
  // -------------------------------------------------------------------

  /**
   * Initialize the protocol session.
   *
   * For CDP (Chrome/Servo/LightPanda): discovers page targets, attaches to one
   * with flatten:true to get a sessionId, then enables Page + Runtime domains
   * on that target. This is required because the server proxies to Chrome's
   * browser-level CDP endpoint.
   *
   * For BiDi (Firefox): gets or creates a browsing context.
   *
   * For "auto" mode: tries CDP first (Target.setDiscoverTargets). If it fails
   * (e.g. we actually got a BiDi session), falls back to BiDi.
   */
  async init(): Promise<void> {
    if (this.protocol === 'auto') {
      await this.autoDetectAndInit();
      return;
    }

    if (this.cdp) {
      await this.cdp.attachToPage();
    } else if (this.bidi) {
      await this.bidi.getOrCreateContext();
    }
  }

  /**
   * Auto-detect protocol by trying CDP first, falling back to BiDi.
   * Used when browser type is "auto" and we don't know what the server gave us.
   */
  private async autoDetectAndInit(): Promise<void> {
    const sessionOpts = this.adapterOpts.commandTimeoutMs
      ? { commandTimeoutMs: this.adapterOpts.commandTimeoutMs }
      : undefined;

    // Try CDP first (most common — Chrome is the default)
    try {
      this.cdp = new CDPSession(this.transport, sessionOpts);
      // Re-wire message routing to include CDP
      this.transport.onMessage((data: string) => this.routeMessage(data));
      await this.cdp.attachToPage();
      this.protocol = 'cdp';
      logger.info('auto-detected CDP protocol');
      return;
    } catch {
      // CDP failed — likely a BiDi session
      this.cdp?.destroy();
      this.cdp = null;
    }

    // Try BiDi
    this.bidi = new BiDiSession(this.transport, sessionOpts);
    this.transport.onMessage((data: string) => this.routeMessage(data));
    await this.bidi.getOrCreateContext();
    this.protocol = 'bidi';
    logger.info('auto-detected BiDi protocol');
  }

  /** Navigate to URL. */
  async navigate(url: string): Promise<void> {
    if (this.cdp) {
      await this.cdp.navigate(url);
    } else {
      await this.bidi!.navigate(url);
    }
  }

  /** Navigate without waiting for full load — for use with polling content extraction. */
  async navigateFast(url: string): Promise<void> {
    if (this.cdp) {
      await this.cdp.navigateFast(url);
    } else {
      await this.bidi!.navigate(url); // BiDi doesn't support fast nav, use regular
    }
  }

  /** Navigate and return on DOMContentLoaded (3s max) — fastest option for SPAs. */
  async navigateDom(url: string): Promise<void> {
    if (this.cdp) {
      await this.cdp.navigateDom(url);
    } else {
      await this.bidi!.navigate(url); // BiDi doesn't support DOM nav, use regular
    }
  }

  /** Get page HTML. */
  async getHTML(): Promise<string> {
    if (this.cdp) {
      return this.cdp.getHTML();
    }
    return this.bidi!.getHTML();
  }

  /** Evaluate JavaScript expression. */
  async evaluate(expression: string): Promise<unknown> {
    if (this.cdp) {
      return this.cdp.evaluate(expression);
    }
    return this.bidi!.evaluate(expression);
  }

  /** Capture screenshot as base64 PNG. */
  async captureScreenshot(): Promise<string> {
    if (this.cdp) {
      return this.cdp.captureScreenshot();
    }
    return this.bidi!.captureScreenshot();
  }

  /** Click at viewport coordinates. */
  async clickPoint(x: number, y: number): Promise<void> {
    if (this.cdp) {
      await this.cdp.clickPoint(x, y);
    } else {
      await this.bidi!.clickPoint(x, y);
    }
  }

  /** Right-click at coordinates. */
  async rightClickPoint(x: number, y: number): Promise<void> {
    if (this.cdp) {
      await this.cdp.rightClickPoint(x, y);
    } else {
      // BiDi right-click via pointer actions
      await this.bidi!.performActions([
        {
          type: 'pointer',
          id: 'mouse',
          actions: [
            { type: 'pointerMove', x: Math.round(x), y: Math.round(y) },
            { type: 'pointerDown', button: 2 },
            { type: 'pointerUp', button: 2 },
          ],
        },
      ]);
    }
  }

  /** Double-click at coordinates. */
  async doubleClickPoint(x: number, y: number): Promise<void> {
    if (this.cdp) {
      await this.cdp.doubleClickPoint(x, y);
    } else {
      await this.bidi!.performActions([
        {
          type: 'pointer',
          id: 'mouse',
          actions: [
            { type: 'pointerMove', x: Math.round(x), y: Math.round(y) },
            { type: 'pointerDown', button: 0 },
            { type: 'pointerUp', button: 0 },
            { type: 'pointerDown', button: 0 },
            { type: 'pointerUp', button: 0 },
          ],
        },
      ]);
    }
  }

  /** Click and hold at coordinates. */
  async clickHoldPoint(x: number, y: number, holdMs: number): Promise<void> {
    if (this.cdp) {
      await this.cdp.clickHoldPoint(x, y, holdMs);
    } else {
      await this.bidi!.performActions([
        {
          type: 'pointer',
          id: 'mouse',
          actions: [
            { type: 'pointerMove', x: Math.round(x), y: Math.round(y) },
            { type: 'pointerDown', button: 0 },
            { type: 'pause', duration: holdMs },
            { type: 'pointerUp', button: 0 },
          ],
        },
      ]);
    }
  }

  /** Hover at coordinates. */
  async hoverPoint(x: number, y: number): Promise<void> {
    if (this.cdp) {
      await this.cdp.hoverPoint(x, y);
    } else {
      await this.bidi!.performActions([
        {
          type: 'pointer',
          id: 'mouse',
          actions: [{ type: 'pointerMove', x: Math.round(x), y: Math.round(y) }],
        },
      ]);
    }
  }

  /** Smooth drag from point to point. */
  async dragPoint(fromX: number, fromY: number, toX: number, toY: number): Promise<void> {
    if (this.cdp) {
      await this.cdp.dragPoint(fromX, fromY, toX, toY);
    } else {
      const steps = 10;
      const actions: any[] = [
        { type: 'pointerMove', x: Math.round(fromX), y: Math.round(fromY) },
        { type: 'pointerDown', button: 0 },
      ];
      for (let i = 1; i <= steps; i++) {
        const t = i / steps;
        actions.push({
          type: 'pointerMove',
          x: Math.round(fromX + (toX - fromX) * t),
          y: Math.round(fromY + (toY - fromY) * t),
          duration: 16,
        });
      }
      actions.push({ type: 'pointerUp', button: 0 });
      await this.bidi!.performActions([{ type: 'pointer', id: 'mouse', actions }]);
    }
  }

  /** Insert text. */
  async insertText(text: string): Promise<void> {
    if (this.cdp) {
      await this.cdp.insertText(text);
    } else {
      await this.bidi!.insertText(text);
    }
  }

  /** Press a named key (e.g. "Enter", "Tab"). */
  async pressKey(keyName: string): Promise<void> {
    const { key, code, keyCode } = getKeyParams(keyName);
    if (this.cdp) {
      await this.cdp.pressKey(key, code, keyCode);
    } else {
      await this.bidi!.performActions([
        {
          type: 'key',
          id: 'keyboard',
          actions: [
            { type: 'keyDown', value: key },
            { type: 'keyUp', value: key },
          ],
        },
      ]);
    }
  }

  /** Send keyDown. */
  async keyDown(keyName: string): Promise<void> {
    const { key, code, keyCode } = getKeyParams(keyName);
    if (this.cdp) {
      await this.cdp.keyDown(key, code, keyCode);
    } else {
      await this.bidi!.performActions([
        { type: 'key', id: 'keyboard', actions: [{ type: 'keyDown', value: key }] },
      ]);
    }
  }

  /** Send keyUp. */
  async keyUp(keyName: string): Promise<void> {
    const { key, code, keyCode } = getKeyParams(keyName);
    if (this.cdp) {
      await this.cdp.keyUp(key, code, keyCode);
    } else {
      await this.bidi!.performActions([
        { type: 'key', id: 'keyboard', actions: [{ type: 'keyUp', value: key }] },
      ]);
    }
  }

  /** Set viewport dimensions. */
  async setViewport(
    width: number,
    height: number,
    deviceScaleFactor: number = 2,
    mobile: boolean = false,
  ): Promise<void> {
    if (this.cdp) {
      await this.cdp.setViewport(width, height, deviceScaleFactor, mobile);
    } else {
      // BiDi doesn't have a direct viewport override — use script
      await this.bidi!.evaluate(
        `window.resizeTo(${width}, ${height})`,
      );
    }
  }

  /** Subscribe to a CDP domain event (only relevant for CDP). */
  onProtocolEvent(method: string, handler: (params: Record<string, unknown>) => void): void {
    if (this.cdp) {
      this.cdp.on(method, handler);
    } else if (this.bidi) {
      this.bidi.on(method, handler);
    }
  }

  /** Clean up resources. Nulls references first to prevent stale message routing. */
  destroy(): void {
    const cdp = this.cdp;
    const bidi = this.bidi;
    // Null out first so routeMessage() drops any in-flight messages during destroy
    this.cdp = null;
    this.bidi = null;
    cdp?.destroy();
    bidi?.destroy();
  }
}

File: rust/src/protocol/protocol_adapter.rs
//! Unified protocol adapter — wraps either CDPSession or BiDiSession.

use crate::errors::{Result, SpiderError};
use crate::events::SpiderEventEmitter;
use crate::protocol::bidi_session::BiDiSession;
use crate::protocol::cdp_session::CDPSession;
use crate::protocol::types::get_key_params;
use serde_json::{json, Value};
use std::sync::Arc;
use tokio::sync::mpsc;
use tracing::{debug, info};

/// Protocol adapter options.
pub struct ProtocolAdapterOptions {
    pub command_timeout_ms: Option<u64>,
}

/// Determines which protocol to use.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ProtocolType {
    Cdp,
    Bidi,
    Auto,
}

/// Unified protocol interface wrapping CDPSession or BiDiSession.
pub struct ProtocolAdapter {
    cdp: Option<CDPSession>,
    bidi: Option<BiDiSession>,
    protocol: ProtocolType,
    send_tx: mpsc::UnboundedSender<String>,
    emitter: SpiderEventEmitter,
    command_timeout_ms: u64,
}

impl ProtocolAdapter {
    pub fn new(
        send_tx: mpsc::UnboundedSender<String>,
        emitter: SpiderEventEmitter,
        browser: &str,
        opts: Option<ProtocolAdapterOptions>,
    ) -> Self {
        let timeout = opts.as_ref()
            .and_then(|o| o.command_timeout_ms)
            .unwrap_or(30_000);

        let (cdp, bidi, protocol) = if browser == "auto" {
            (None, None, ProtocolType::Auto)
        } else if browser == "firefox" {
            (None, Some(BiDiSession::new(send_tx.clone(), timeout)), ProtocolType::Bidi)
        } else {
            (Some(CDPSession::new(send_tx.clone(), timeout)), None, ProtocolType::Cdp)
        };

        Self {
            cdp,
            bidi,
            protocol,
            send_tx,
            emitter,
            command_timeout_ms: timeout,
        }
    }

    pub fn protocol_type(&self) -> ProtocolType {
        self.protocol
    }

    /// Route incoming WebSocket messages to the right session.
    pub fn route_message(&self, data: &str) {
        // Check for Spider.* events first
        if let Ok(msg) = serde_json::from_str::<Value>(data) {
            if let Some(method) = msg.get("method").and_then(|v| v.as_str()) {
                if method.starts_with("Spider.") {
                    self.handle_spider_event(method, msg.get("params").cloned().unwrap_or(json!({})));
                    return;
                }
            }
        }

        if let Some(ref cdp) = self.cdp {
            cdp.handle_message(data);
        } else if let Some(ref bidi) = self.bidi {
            bidi.handle_message(data);
        }
    }

    fn handle_spider_event(&self, method: &str, params: Value) {
        match method {
            "Spider.captchaDetected" => {
                self.emitter.emit("captcha.detected", params);
            }
            "Spider.captchaSolving" => {
                self.emitter.emit("captcha.solving", params);
            }
            "Spider.captchaSolved" => {
                self.emitter.emit("captcha.solved", params);
            }
            "Spider.captchaFailed" => {
                self.emitter.emit("captcha.failed", params);
            }
            _ => {
                debug!("unhandled Spider event: {}", method);
            }
        }
    }

    /// Initialize the protocol session.
    pub async fn init(&mut self) -> Result<()> {
        if self.protocol == ProtocolType::Auto {
            self.auto_detect_and_init().await?;
            return Ok(());
        }

        if let Some(ref cdp) = self.cdp {
            cdp.attach_to_page().await?;
        } else if let Some(ref bidi) = self.bidi {
            bidi.get_or_create_context().await?;
        }
        Ok(())
    }

    async fn auto_detect_and_init(&mut self) -> Result<()> {
        // Try CDP first
        let cdp = CDPSession::new(self.send_tx.clone(), self.command_timeout_ms);
        match cdp.attach_to_page().await {
            Ok(_) => {
                self.cdp = Some(cdp);
                self.protocol = ProtocolType::Cdp;
                info!("auto-detected CDP protocol");
                return Ok(());
            }
            Err(_) => {
                cdp.destroy();
            }
        }

        // Try BiDi
        let bidi = BiDiSession::new(self.send_tx.clone(), self.command_timeout_ms);
        bidi.get_or_create_context().await?;
        self.bidi = Some(bidi);
        self.protocol = ProtocolType::Bidi;
        info!("auto-detected BiDi protocol");
        Ok(())
    }

    // ------------------------------------------------------------------
    // Unified interface
    // ------------------------------------------------------------------

    pub async fn navigate(&self, url: &str) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.navigate(url).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.navigate(url).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn navigate_fast(&self, url: &str) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.navigate_fast(url).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.navigate(url).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn navigate_dom(&self, url: &str) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.navigate_dom(url).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.navigate(url).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn get_html(&self) -> Result<String> {
        if let Some(ref cdp) = self.cdp {
            cdp.get_html().await
        } else if let Some(ref bidi) = self.bidi {
            bidi.get_html().await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn evaluate(&self, expression: &str) -> Result<Value> {
        if let Some(ref cdp) = self.cdp {
            cdp.evaluate(expression).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.evaluate(expression).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn capture_screenshot(&self) -> Result<String> {
        if let Some(ref cdp) = self.cdp {
            cdp.capture_screenshot().await
        } else if let Some(ref bidi) = self.bidi {
            bidi.capture_screenshot().await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn click_point(&self, x: f64, y: f64) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.click_point(x, y).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.click_point(x, y).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn right_click_point(&self, x: f64, y: f64) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.right_click_point(x, y).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "pointer", "id": "mouse",
                "actions": [
                    {"type": "pointerMove", "x": x.round() as i64, "y": y.round() as i64},
                    {"type": "pointerDown", "button": 2},
                    {"type": "pointerUp", "button": 2},
                ]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn double_click_point(&self, x: f64, y: f64) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.double_click_point(x, y).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "pointer", "id": "mouse",
                "actions": [
                    {"type": "pointerMove", "x": x.round() as i64, "y": y.round() as i64},
                    {"type": "pointerDown", "button": 0},
                    {"type": "pointerUp", "button": 0},
                    {"type": "pointerDown", "button": 0},
                    {"type": "pointerUp", "button": 0},
                ]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn click_hold_point(&self, x: f64, y: f64, hold_ms: u64) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.click_hold_point(x, y, hold_ms).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "pointer", "id": "mouse",
                "actions": [
                    {"type": "pointerMove", "x": x.round() as i64, "y": y.round() as i64},
                    {"type": "pointerDown", "button": 0},
                    {"type": "pause", "duration": hold_ms},
                    {"type": "pointerUp", "button": 0},
                ]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn hover_point(&self, x: f64, y: f64) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.hover_point(x, y).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "pointer", "id": "mouse",
                "actions": [{"type": "pointerMove", "x": x.round() as i64, "y": y.round() as i64}]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn drag_point(&self, from_x: f64, from_y: f64, to_x: f64, to_y: f64) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.drag_point(from_x, from_y, to_x, to_y).await
        } else if let Some(ref bidi) = self.bidi {
            let steps = 10;
            let mut actions = vec![
                json!({"type": "pointerMove", "x": from_x.round() as i64, "y": from_y.round() as i64}),
                json!({"type": "pointerDown", "button": 0}),
            ];
            for i in 1..=steps {
                let t = i as f64 / steps as f64;
                actions.push(json!({
                    "type": "pointerMove",
                    "x": (from_x + (to_x - from_x) * t).round() as i64,
                    "y": (from_y + (to_y - from_y) * t).round() as i64,
                    "duration": 16,
                }));
            }
            actions.push(json!({"type": "pointerUp", "button": 0}));
            bidi.perform_actions(json!([{"type": "pointer", "id": "mouse", "actions": actions}])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn insert_text(&self, text: &str) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.insert_text(text).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.insert_text(text).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn press_key(&self, key_name: &str) -> Result<()> {
        let (key, code, key_code) = get_key_params(key_name);
        if let Some(ref cdp) = self.cdp {
            cdp.press_key(key, code, key_code).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "key", "id": "keyboard",
                "actions": [
                    {"type": "keyDown", "value": key},
                    {"type": "keyUp", "value": key},
                ]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn key_down(&self, key_name: &str) -> Result<()> {
        let (key, code, key_code) = get_key_params(key_name);
        if let Some(ref cdp) = self.cdp {
            cdp.key_down(key, code, key_code).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "key", "id": "keyboard",
                "actions": [{"type": "keyDown", "value": key}]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn key_up(&self, key_name: &str) -> Result<()> {
        let (key, code, key_code) = get_key_params(key_name);
        if let Some(ref cdp) = self.cdp {
            cdp.key_up(key, code, key_code).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.perform_actions(json!([{
                "type": "key", "id": "keyboard",
                "actions": [{"type": "keyUp", "value": key}]
            }])).await
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub async fn set_viewport(&self, width: u32, height: u32, dpr: f64, mobile: bool) -> Result<()> {
        if let Some(ref cdp) = self.cdp {
            cdp.set_viewport(width, height, dpr, mobile).await
        } else if let Some(ref bidi) = self.bidi {
            bidi.evaluate(&format!("window.resizeTo({width}, {height})")).await?;
            Ok(())
        } else {
            Err(SpiderError::Protocol("No protocol session".into()))
        }
    }

    pub fn on_protocol_event(&self, method: &str, handler: Arc<dyn Fn(Value) + Send + Sync>) {
        if let Some(ref cdp) = self.cdp {
            cdp.on(method, handler);
        } else if let Some(ref bidi) = self.bidi {
            bidi.on(method, handler);
        }
    }

    pub fn destroy(&self) {
        if let Some(ref cdp) = self.cdp {
            cdp.destroy();
        }
        if let Some(ref bidi) = self.bidi {
            bidi.destroy();
        }
    }
}
